packages/teleport/src/DesktopSession/useDesktopSession.tsx

```diff
@@ -59,44 +59,67 @@ export default function useDesktopSession(ctx: Ctx) {
       .replace(':desktopId', desktopId)
       .replace(':token', getAccessToken());
 
-    const cli = new TdpClient(addr, username);
-
-    // If the websocket is closed remove all listeners that depend on it.
-    // If it was closed intentionally by the user, set attempt to disconnected,
-    // otherwise assume a server error.
-    cli.on('close', (message: { userDisconnected: boolean }) => {
-      if (message.userDisconnected) {
-        setAttempt({ status: 'disconnected' });
-      } else {
-        setAttempt({
-          status: 'failed',
-          statusText: 'server error',
-        });
-      }
-      cli.removeAllListeners();
-    });
-
-    cli.on('error', () => {
-      setAttempt({
-        status: 'failed',
-        statusText: 'connection error',
-      });
-    });
-
-    return cli;
+    return new TdpClient(addr, username);
   }, [clusterId, username, desktopId]);
 
   useEffect(() => {
     Promise.all([
       ctx.desktopService.fetchDesktops(clusterId),
-      tdpClient.connect(),
+      // This Promise effectively takes the events emmitted during the tdpClient's
+      // initial attempted connection to the websocket and converts them into a standard
+      // promise that resolves on success and rejects on failure.
+      new Promise<void>((resolve, reject) => {
+        tdpClient.on('open', () => {
+          resolve();
+        });
+
+        tdpClient.on('error', (err: Error) => {
+          reject(err);
+        });
+
+        tdpClient.connect();
+      }),
     ])
       .then(vals => {
         makeUserHost(vals[0]);
+
+        // Remove the listeners from the Promise.all call and then initialize
+        // many of the listeners that will last the client's lifetime.
+        tdpClient.removeAllListeners('open');
+        tdpClient.removeAllListeners('error');
+
+        // Triggered in the event of a websocket protocol error.
+        tdpClient.on('error', () => {
+          setAttempt({
+            status: 'failed',
+            statusText: 'websocket protocol error',
+          });
+
+          tdpClient.removeAllListeners();
+        });
```

Comment:
looks complicated, I wonder why did not you like the previous version where tdpClient was close to `canvas` element?

Reply:
I think you're referring to [this version here](https://github.com/gravitational/webapps/blob/222768dd9bb95843c1950032965948aa79a45c22/packages/teleport/src/DesktopSession/DesktopSession.tsx#L44-L88). The original impetus for moving away from that design was Lisa's comment [here](https://github.com/gravitational/webapps/pull/391#discussion_r707970871), where she suggested keeping state management encapsulated in the `useDesktopSession` hook (my sense is this is standard practice in the codebase with infrequent exceptions). 

This implementation does that, while at the same time implementing Lisa's other suggestion [here](https://github.com/gravitational/webapps/pull/391#discussion_r708007444). With this, we only need to track a single `DesktopSessionAttempt` for the entire connection state, rather than tracking both `wsAttempt` and `fetchDesktopAttempt`, and use [this gross and complex logic to synchronize them](https://github.com/gravitational/webapps/blob/222768dd9bb95843c1950032965948aa79a45c22/packages/teleport/src/DesktopSession/useDesktopSession.tsx#L81-L99).

I agree though, that this is complicated. Another approach lies with making `tdpClient.connect()` [return a `Promise`](https://github.com/gravitational/webapps/blob/2790402525dcfb43a73eb7789004ae151e752c70/packages/teleport/src/lib/tdp/client.ts#L35-L69). Then, this big disgusting thing can just be
```typescript
  useEffect(() => {
    Promise.all([
      ctx.desktopService.fetchDesktops(clusterId),
      tdpClient.connect(),
    ])
      .then(vals => {
        makeUserHost(vals[0]);
        setAttempt({ status: 'success' });
      })
      .catch(err => {
        setAttempt({ status: 'failed', statusText: err.message });
      });
  }, [clusterId, username, desktopId]);
```

(I had this in a previous commit, and then created this mess based on what I now see as a mistaken premise that it was important to keep the `tdpClient` an `EventEmitter` exclusively, and not confuse a user by giving them the option to deal with a `Promise` _or_ an `'open'` Event. That doesn't hold up to scrutiny though, since nobody is confused...)

In that case, we have the option of either adding these event listeners here or in `DesktopSession` next to the `canvas`. My argument for keeping them in here is that by doing so, `tdpClient` events that create side effects by changing state are encapsulated in the hook, alongside any api calls (currently `ctx.desktopService.fetchDesktops(clusterId)`) and the initial websocket connection which do the same. Only the `'render`' event listener need be in `DesktopSession`, distinguished by the fact that it doesn't change the component's state and requires the `canvas` to draw to.

Reply:
I think we are dealing with 2 states: one is the session and another one is the canvas. Sometimes it's OK to lift a state up and sometimes it does not. In this case I would keep it closer to the `canvas` as it makes it easier to reason about (because tdpClient _updates_ canvas, canvas is the state and tdpClient is an API).